runtime/compiler/env/DependencyTable.cpp

```diff
@@ -32,9 +33,126 @@ TR_AOTDependencyTable::TR_AOTDependencyTable(TR_J9SharedCache *sharedCache) :
    _isActive(true),
    _sharedCache(sharedCache),
    _tableMonitor(TR::Monitor::create("JIT-AOTDependencyTableMonitor")),
-   _offsetMap(decltype(_offsetMap)::allocator_type(TR::Compiler->persistentAllocator()))
+   _offsetMap(decltype(_offsetMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _methodMap(decltype(_methodMap)::allocator_type(TR::Compiler->persistentAllocator())),
+   _pendingLoads(decltype(_pendingLoads)::allocator_type(TR::Compiler->persistentAllocator()))
    { }
 
+bool
+TR_AOTDependencyTable::trackMethod(J9VMThread *vmThread, J9Method *method, J9ROMMethod *romMethod, bool &dependenciesSatisfied)
+   {
+   const uintptr_t *methodDependencies = NULL;
+   if (!_sharedCache->methodHasAOTBodyWithDependencies(vmThread, romMethod, methodDependencies))
+      return false;
+
+   if (!methodDependencies)
+      {
+      dependenciesSatisfied = true;
+      return true;
+      }
+
+   OMR::CriticalSection cs(_tableMonitor);
+   if (!isActive())
+      return false;
+
+   try
+      {
+      uintptr_t totalDependencies = *methodDependencies;
+      uintptr_t numberRemainingDependencies = totalDependencies;
+
+      auto m_it = _methodMap.insert({method, {0, methodDependencies}});
+      auto methodEntry = &(*m_it.first);
+
+      for (size_t i = 1; i <= totalDependencies; ++i)
+         {
+         bool needsInitialization = false;
+         uintptr_t chainOffset = decodeDependencyOffset(methodDependencies[i], needsInitialization);
+         uintptr_t offset = _sharedCache->startingROMClassOffsetOfClassChain(_sharedCache->pointerFromOffsetInSharedCache(chainOffset));
+         auto entry = getOffsetEntry(offset, true);
+         if (needsInitialization)
```

Comment:
I had it set up this way because I wanted to be able to "unsatisfy" the dependencies on class unload or redefinition. That means that `methodEntry` has to be included in the `offsetEntry` of every dependency as long as it is being tracked. So I could skip adding `methodEntry` to `offsetEntry` here, but then I'd have to go back and add it in at the very end if any dependencies remained.

Reply:
So, you are saying that `_waitingInitMethods` and `_waitingLoadMethods` at any given time contain all of the original dependencies. When some of the dependencies get satisfied, we never delete from these lists. The list is only deleted when the method is no longer tracked because it has been compiled. Is that correct?

Reply:
That's correct, though technically we do `stopTracking()` once all of the dependencies are satisfied, which happens before compilation.